Avoid not disappearing contribution annotations

m273d15 opened the pull request:
The PR splits all annotations in annotations of size 1 and removes the old splitting logic. This is much easier and fixes the issue. However, I would assume that this was the initial solution in Saros and the split logic was an optimization, but during my tests (manual and with stf) I had not the impression that my change worsens the user experiences.

Furthermore, this slightly changes the history handling, because text insertions are multiple annotation insertions. Therefore a long insertion is only annotated at the end. This leads to the question "How should this feature work?".

@srossbach What are your thoughts? Did I miss a corner case that requires the other handling?

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  -2
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)

srossbach commented:
That is the big question. How should it work ? If a C&P should be fully annotated then you current solution could be a real performance nightmare. Adding and removing annotations is more CPU intensive then playing a modern computer game. Although we are already optimized it in such way that activities are executed all at once on the EDT if the CPU is not fast enough. If you CPU is fast enough CPU load goes up to 100% of one CPU die.

srossbach commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -124,9 +124,17 @@ public void insertAnnotation(IAnnotationModel model, int offset, int length, Use
 
-    if (length < 0) // why is 0 len allowed ?
-    return;
+    for (int i = 0; i < length; i++) {
```
My big guess is that if you C&P a long enough text, the receiving side will freeze.

m273d15 replied:
What do you think about checking `length > MAX_HISTORY_LENGTH` and just annotating the the last `MAX_HISTORY_LENGTH`s characters? Is a simple optimization that reduces the issue.

srossbach replied:
@tobous How are annotations handled in IntelliJ regarding C&P, e.g one insert = 50 characters ?

And @m273d15 if you want to make mass updates in the annotation model, please calculate the added and removed annotations before you actually replace them. We have a helper function that tries to that at once, i.e only making 1 update call for all affected annotations.

tobous replied:
In IntelliJ, contribution annotations are handled similarly but not exactly in the way implemented here.

The base approach is the same: All contribution annotations are split into one character wide annotations that do not auto-extend (i.e. don't extend when text is added at either end). See [AnnotationManager.addContributionAnnotation(...)](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/AnnotationManager.java#L120).
I made this decision to avoid such cases where another user types in the middle of an annotation, thereby extending it if it were a continuously annotated block.

Where the handling seems to differ is that all text added by one activity (like a C&P) is still seen as one annotation, even though it is displayed in many different parts. See [ContributionAnnotation](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/ContributionAnnotation.java). One annotation has a list of `AnnotationRange` objects, which each represent a highlighted section of text.
As a result, all highlightings for the action are bundled and seen as one annotation for the annotation rotation.

In IntelliJ, the performance for this is ok. (I haven't measured the impact.) The funny thing is that, even without Saros, C&P a large chunk (~2000 line) in a VERY large file (~250000 lines) freezes the local editor. Such issues, however, do not occur on the receiving side of such activities. As a result, during a running session, the remove site (which has to deal with all of the annotations but still runs without a hitch) is done far before the sending site.

srossbach replied:
Ok so the logic what is displayed is the same as in Eclipse. I do not see a problem by just using single annotations in Eclipse too. But as I already said, use the bulk method.

srossbach replied:
My last 2 cents. It should not be that hard to modify the history in such way that it stores multiples annotations.

```
List annotationsToRemove = history.getAndRemoveAnnotationsForOldestEntry();
List annotationsToAdd = createAnnotationsForContributionRange(....)
history.addEntry(annotationsToAdd)
model.replaceAnnotations(annotationsToRemove, annotationsToAdd)
```

m273d15 replied:
I will modify the history handling and adopt the `AnnotationModelHelper.replaceAnnotationsInModel` method (I think this is the bulk method you talked about).

srossbach replied:
The complexity of the algorithm went up from O(m) to O(m*n). m = number of annotations in the model, n = characters / annotations to add.

Furthermore is it that tragic if multiple annotations are present for a single position ? (In reality it should not happen at all, but it does see the issue in the EditorManager)

As you already know the covered range and if you want to keep the check this algorithm can be optimized to have an average runtime of O(m) again, e.g calculate the range, traverse the annotations only once and collect all that do collide, calculate the colliding annotations afterwards.

tobous commented on eclipse/src/saros/editor/EditorManager.java:
```diff
@@ -1079,12 +1071,11 @@ private void replaceText(SPath path, int offset, String replacedText, String tex
         }
       }
 
-      IAnnotationModel model = provider.getAnnotationModel(input);
```
Do you know why there was a separate handling beforehand? I.e. why is this no longer necessary?

srossbach replied:
Nobody really knows that. As I already said the logic is buggy, annotations are inserted twice (the contribution manager silently ignores it).

Guess it makes more sense either to remove this line or count the editors in the loop above and only execute this code here if you annotated 0 editors.

srossbach replied:
And I guess nobody know if annotation removal from the history is correctly working when the annotation is used in multiple editors (e.g splitview)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity decreasing per file
==============================
+ eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  -2
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)

srossbach commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -122,73 +108,72 @@ public ContributionAnnotationManager(
   /**
-   * Splits the contribution annotation at given position, so that a following text change will not
-   * expand the annotation. This needs to be called before the text is changed.
+   * Creates one contribution annotations with length 1 for each char contained in the range defined
```
Creates contribution annotations

srossbach commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -122,73 +108,72 @@ public ContributionAnnotationManager(
-   * Splits the contribution annotation at given position, so that a following text change will not
-   * expand the annotation. This needs to be called before the text is changed.
+   * Creates one contribution annotations with length 1 for each char contained in the range defined
+   * by {@code offset} and {@code length}. The annotation is inserted to given model if there is not
```
This method only performs the calculation, it does not alter the model.

srossbach commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -122,73 +108,72 @@ public ContributionAnnotationManager(
+   */
+  private Pair<ContributionAnnotation, Position> createPositionedAnnotation(
+      IAnnotationModel model, int offset, User source) {
+    final int ANNOTATION_SIZE = 1;
```
ANNOTATION_LENGTH ?

srossbach commented on eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java:
```diff
@@ -232,19 +155,49 @@ public void testRemoveAnnotationsWhenUserLeaves() {
   }
 
+  @Test
+  public void testInsertAnnotationWithLengthGreaterOne() {
```
I do not know if this is sufficient or to just introduce a test that combines multiple annotations lengths. Currently all other tests are only working with a length of 1.

m273d15 replied:
I added the test "testHistoryRemovelWithDifferentLengths"

srossbach commented on eclipse/src/saros/editor/EditorManager.java:
```diff
@@ -633,14 +633,6 @@ void textAboutToBeChanged(int offset, String text, int replaceLength, IDocument
-    IEditorInput input = changedEditor.getEditorInput();
-    IDocumentProvider provider = EditorAPI.getDocumentProvider(input);
-    IAnnotationModel model = provider.getAnnotationModel(input);
-    contributionAnnotationManager.splitAnnotation(model, offset);
```
Can you please have a look at the conversation history ? I already talked about another issue in this file on line 1086.

srossbach commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -389,4 +327,69 @@ private void removeFromHistoryAndAnnotationModel(final Collection<History> histo
                   && users.contains(((ContributionAnnotation) (a)).getSource()));
   }
+
+  static final class History {
```
Any reason to not move the code to a separate class ?

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  2
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java  2
- eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  1
- eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java  5
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)

m273d15 commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -122,73 +108,72 @@ public ContributionAnnotationManager(
-
-    if (length < 0) // why is 0 len allowed ?
-    return;
+    if (!contribtionAnnotationsEnabled || length < 0) return;
```
insertAnnotation is called by EditorManager#replaceText, which is also called if a deletion of a char is performed. This leads to length == 0.I would like to change the handling, but I had to stop with refactoring and rework because the PR would be too large.

I restructured the code in order to make the cases more clear.

m273d15 commented:
@srossbach I was not able to directly answer these comments:

> Guess it makes more sense either to remove this line or count the editors in the loop above and only execute this code here if you annotated 0 editors.

I removed the additional insert of annotation in the EditorManager and removed the check in the ContributionAnnotationManager.

> And I guess nobody know if annotation removal from the history is correctly working when the annotation is used in multiple editors (e.g splitview)

I fixed the splitview handling. The part that removed the old annotation from the model used the wrong model.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  1
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java  2
- eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java  4
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  2
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)

tobous commented on eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java:
```diff
@@ -300,4 +357,24 @@ private int getAnnotationCount(IAnnotationModel model) {
     return positions;
   }
+
+  private User createAliceTestUser() {
```
As the returned object is always the same, these methods could also be replaced by a static variable for each user. Or are the user objects modified by the tests?

srossbach commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -112,83 +97,99 @@ public ContributionAnnotationManager(
+   * @param model to add the annotation to.
+   * @param offset start of the annotation to add.
+   * @param source of the annotation.
+   * @returns a pair containing the annotation and its position or <code>null</code> if an
```
The method no longer returns null because the collision check was removed.

srossbach commented on eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java:
```diff
@@ -0,0 +1,100 @@
+   * [[A, B], [C], [D, E], [F], [G]] -> [[A, B], [C], [D', E], [F], [G]]
+   *
+   */
+  protected void replaceInHistory(
```
No longer used ?

srossbach commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -367,18 +271,17 @@ private void removeAnnotationsForUser(final User user) {
-        annotationModels.add(entry.annotation.getModel());
-        users.add(entry.annotation.getSource());
+    for (final ContributionAnnotationHistory history : histories) {
+      List<ContributionAnnotation> allAnnotationsInHistory = history.clear();
```
Just my 2 cents: Either use the final key constantly or do not use it at all. The code looks pretty weird and you can already see without using GIT Blame that it was written by at least two different developers.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java  2
- eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java  4
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  -2
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)

m273d15 commented:
@tobous Based on your comments, I changed the "insertAnnotation" method to directly return if length <= 0. With this change it was possible to merge the "removeEntries()" and "addNewEntry()" method and I was able to refactor the history again. Furthermore, I changed the tests based on your comments.

I added all missing final keyword automatically with IntelliJ.

tobous commented on eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  @Test
+  public void testaddEntriesExceedingMaxLength() {
```
```suggestion
  public void testAddEntriesExceedingMaxLength() {
```

tobous commented on eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java:
```diff
@@ -0,0 +1,86 @@
+    assertEquals(MAX_HISTORY_LENGTH, history.getSize());
+
+    annotationsToRemove = history.addNewEntry(createDummyAnnotation(testModel));
+    assertThat(annotationsToRemove.getRight(), is(firstEntry));
```
Ok, this does work as `ContributionAnnotation` does not implement an equals method, meaning all comparisons are done based on object references anyways. But this could be confusing when first reading the test. If it were implementing equals, the test would not ensure the correct functionality as all dummy annotations contain the same content, meaning they would be seen as equal.

So this adjustment is not strictly necessary, but I would still suggest explicitly comparing the object references using `assertSame(...)` just to avoid such confusion (or adding a comment explaining the situation).

tobous commented on eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java:
```diff
@@ -0,0 +1,86 @@
+    assertEquals(testModel2, annotationsToRemove.getLeft());
+
+    annotationsToRemove = history.addNewEntry(createDummyAnnotation(testModel));
+    assertEquals(testModel, annotationsToRemove.getLeft());
```
Could also use `assertSame(...)`.

m273d15 commented on eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java:
```diff
@@ -85,16 +76,35 @@ protected void replaceInHistory(
+    if (queue.size() > maxHistoryLength) {
       throw new IllegalStateException(
-          "The queue already contains the " + "allowed number of annotations.");
+          "The queue already contains the allowed number of annotations.");
```
I don't know why I included the check. I wanted to remove the if block.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  2
- eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java  4
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  -2
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java  4
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  2
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  -2
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java  2
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  2
- eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java  4
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  -1
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)

m273d15 commented:
@tobous I rebased and added [this commit](https://github.com/saros-project/saros/pull/866/commits/0d13ed7b831139a2b959959fc9b9f81a01277ac7) in order to avoid the corner case:
* paste at a position where an annotation exists
* the IDocument#replace method resizes the corresponding annotations
The new commit removes the resized annotation.

tobous commented on eclipse/src/saros/editor/EditorManager.java:
```diff
@@ -1062,6 +1054,8 @@ private void replaceText(SPath path, int offset, String replacedText, String tex
       // Try to replace
       try {
+        // Attention: This method also alters the annotation model if an
+        // annotation exists at the offset and the length &gt; 1
```
Use correct symbol instead of character code?
```suggestion
        // annotation exists at the offset and the length > 1
```

tobous commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -33,32 +34,20 @@
  * they belong to are deleted, the session ends, or their respective author leaves the session. To
  * avoid cluttering the editors, only the last {@value #MAX_HISTORY_LENGTH} changes are annotated.
  *
+ * <p>The invariant of the class should be that all handled annotations have a length &lt;= 1. This
```
```suggestion
 * <p>The invariant of the class should be that all handled annotations have a length <= 1. This
```

tobous commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -112,102 +101,151 @@ public ContributionAnnotationManager(
-    return;
+    if (length > 1) {
+      // Correct current model state to match with the assumption that all
+      // annotations have length &lt;= 1.
```
```suggestion
      // annotations have length <= 1.
```

tobous commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -112,102 +101,151 @@ public ContributionAnnotationManager(
   /**
-   * Splits the contribution annotation at given position, so that a following text change will not
-   * expand the annotation. This needs to be called before the text is changed.
+   * This method is only called if an annotation with length &gt; 1 should be inserted, because if
```
In actual javadoc, this style of using `<` and `>` technically works when viewed through a tool interpreting the javadoc but is still unnecessarily hard to read when looking at the code.
```suggestion
   * This method is only called if an annotation with length > 1 should be inserted, because if
```

m273d15 replied:
Hmm, I use this, because the javadoc processor complains about using "<"/">". I would like to have a javadoc build without multiple errors in the future.

tobous replied:
I guess then keep it the way it is for the javadoc. The normal comments should still be adjusted in my opinion. And in the javadoc, you could consider writing it out instead. But that is a matter of taste.

tobous commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -112,102 +101,151 @@ public ContributionAnnotationManager(
+   * {@link IDocument} that is performed by the caller of {@link #insertAnnotation} leads to an
+   * internal change of the annotation model.
+   *
+   * <p>The annotation with length &gt; 1 is removed, because the new annotation is subsequently
```
```suggestion
   * <p>The annotation with length > 1 is removed, because the new annotation is subsequently
```

tobous commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -112,102 +101,151 @@ public ContributionAnnotationManager(
-   * @param model to search for annotations to split
-   * @param offset at which annotations should be split
+   * <p>Therefore, this methods works against the internal mechanisms of eclipse in order to enforce
+   * the invariant of annotations with have a length &lt;= 1.
```
```suggestion
   * the invariant of annotations with have a length <= 1.
```

tobous commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -112,102 +101,151 @@ public ContributionAnnotationManager(
+        annotationsToRemove.add(contrAnnotation);
 
-      if ((offset > pos.offset) && (offset < pos.offset + pos.length)) {
+        // Expect only one annotation with length &gt; 1 created by
```
```suggestion
        // Expect only one annotation with length > 1 created by
```

tobous commented on eclipse/src/saros/editor/internal/ContributionAnnotationManager.java:
```diff
@@ -112,102 +101,151 @@ public ContributionAnnotationManager(
+   * @param source of the annotation.
+   * @return a map containing the annotations and their positions
+   */
+  private Map<ContributionAnnotation, Position> createAnnotationsForContributionRange(
```
So the annotations are present twice in this case, meaning you can just remove the merged one?

Is there currently any logic re-adding existing annotations to the model? If so, can this become an issue in such cases as well (i.e. the internal optimization to merge the annotations is applied again)?

In general, does this have a noticeable impact on performance?

m273d15 replied:
I found this optimization only if the IDocument#replace method is used. Therefore, I assume that this is not an optimization of the model, but the way the replace method introduces annotations. I also tested the re-adding (e.g. if the use color changes).

With using the optimized replaceAnnotations method I noticed no impact, without the optimized version you notice an impact if you paste long text (the annotations are painted bit by bit). Therefore, we use the optimized version.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/editor/internal/ContributionAnnotationHistory.java  4
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  2
- eclipse/test/junit/saros/editor/internal/ContributionAnnotationHistoryTest.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/editor/internal/ContributionAnnotationManager.java  -1
         

Clones removed
==============
+ eclipse/test/junit/saros/editor/internal/ContributionAnnotationManagerTest.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5134601&bid=16871785)
